### PR TITLE
Druid Query Language (DQL)

### PR DESCRIPTION
**Pull request by anskarl**
This is the implementation of Druid query language (DQL). In brief it provides the following:
  - flow-based builders for timeseries, group-by and top-n queries
  - filter expressions
  - having expressions
  - aggregation and post-aggregation expressions

GH-32

The example below demonstrates a group-by query:

```
case class GroupByIsAnonymous(isAnonymous: String, country: String, count: Int)

val query: GroupByQuery = DQL
    .granularity(GranularityType.Day)
    .interval("2011-06-01/2017-06-01")
    .agg(count as "count")
    .where('countryName.isNotNull)
    .groupBy('isAnonymous, 'countryName.extract(UpperExtractionFn()) as "country")
    .having('count > 100 and 'count < 200)
    .limit(10, 'count.desc(DimensionOrderType.numeric))
    .build()
      
val response = query.execute().map(_.list[GroupByIsAnonymous])
val result: List[GroupByIsAnonymous] = Await.result(response, 10.seconds)
```
In [docs/dql.md](https://github.com/anskarl/scruid/blob/feature/dql/docs/dql.md) you can find documentation and examples.

### CODE REVIEW

**Comment by apollotonkosmo** [comment]
+1

**Comment by thelgis** [comment]
+1

**Comment by zaoudis** [comment]
+1

**Comment by bjgbeelen** [comment]
Hi @anskarl thanks for your pull request. Just wanted to let you know that it;s appreciated, but I haven't found the time yet to properly look a it. Hopefully I can have a look next week. 
Thanks for your patience!

**Comment by anskarl** [comment]
@bjgbeelen no worries :)

File: docs/dql.md [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+'dim between("0.0", "10.0")
+
+// "0.0" <= dim <= "10.0"
+'dim between("0.0", "10.0", lowerStrict = true, upperStrict = true)
+
+// "0.0" <= dim < "10.0"
+'dim between("0.0", "10.0", lowerStrict = true, upperStrict = false)
+```
+
+Additionally, only when strings are given to bound operators, you can to specify a the ordering type and
```
Comment (bjgbeelen) [inline docs/dql.md]: I'm having trouble reading this sentence

File: docs/dql.md [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+Arithmetic post-aggregator can be applied to aggregators or other post aggregators and 
+the supported functions are `+`, `-`, `*`, `/`, and `quotient`.
+
+```
+'dim_name + 2
+
+('count / 2) as "halfCount"
+```
+
+In arithmetic post-aggregators may also specify an ordering, which defines the order of resulting values when
```
Comment (bjgbeelen) [inline docs/dql.md]: In arithmetic jpost-aggregators **you** may?

File: docs/dql.md [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+    .agg(count as "ts_count")
+    .build()
+
+val response = query.execute().map(_.list[TimeseriesCount])
+val result: List[TimeseriesCount] = Await.result(response, 10.seconds)
+```
+
+#### Top-N query
+
+The following query computes the Top-5 `countryName` wrt to the aggregation `agg_count`.
```
Comment (bjgbeelen) [inline docs/dql.md]: wrt ~to~ 

or 

with respect to

File: docs/dql.md [hunk 4]
```diff
@@ -0,0 +1,633 @@
...
+    .from("wikiticker")
+    .granularity(GranularityType.Week)
+    .interval("2011-06-01/2017-06-01")
+    .agg(count as "agg_count")
+    .postAgg(('agg_count / 2) as "half_count")
+    .topN('countryName, metric = "agg_count", threshold = 5)
+    .build()
+
+val response = query.execute().map(_.list[PostAggregationAnonymous])
+val result: List[PostAggregationAnonymous] = Await.result(response, 10.seconds)
```
Comment (bjgbeelen) [inline docs/dql.md]: Same comment as in the `README` would apply here

File: docs/dql.md [hunk 5]
```diff
@@ -0,0 +1,633 @@
...
+    .groupBy('isAnonymous, 'countryName.extract(UpperExtractionFn()) as "country")
+    .having('count > 100 and 'count < 200)
+    .limit(10, 'count.desc(DimensionOrderType.numeric))
+    .build()
+      
+val response = query.execute().map(_.list[GroupByIsAnonymous])
+val result: List[GroupByIsAnonymous] = Await.result(response, 10.seconds)      
+```
+
+
```
[the remark this answers is not in the document]
Comment (bjgbeelen) [reply docs/dql.md]: Thanks for extensive and well written documentation!

File: src/main/scala/ing/wbaa/druid/DruidQuery.scala [hunk 6]
```diff
@@ -63,18 +63,55 @@ case class GroupByQuery(
...
+case class OrderByColumnSpec(
+    dimension: String,
+    direction: Direction = Direction.ascending,
+    dimensionOrder: DimensionOrder = DimensionOrder()
+)
+case class DimensionOrder(`type`: DimensionOrderType = DimensionOrderType.lexicographic)
+
+sealed trait Direction extends Enum with LowerCaseEnumStringEncoder
+object Direction extends EnumCodec[Direction] {
+  case object ascending  extends Direction
```
Comment (bjgbeelen) [inline src/main/scala/ing/wbaa/druid/DruidQuery.scala]: Is there a specific reason to not use a capital in defining the object? That would be `Scala` convention right?. Json serialization would take care of the lowercasing in this case

Comment (bjgbeelen) [reply src/main/scala/ing/wbaa/druid/DruidQuery.scala]: I think there are other places where the same comment applies

File: README.md [hunk 7]
```diff
@@ -75,6 +75,29 @@ val series: Map[ZonedDateTime, TimeseriesCount] = response.series[TimeseriesCoun
...
+    .interval("2011-06-01/2017-06-01")
+    .agg(count as "count")
+    .where('countryName.isNotNull)
+    .groupBy('isAnonymous, 'countryName.extract(UpperExtractionFn()) as "country")
+    .having('count > 100 and 'count < 200)
+    .limit(10, 'count.desc(DimensionOrderType.numeric))
+    .build()
+      
+val response = query.execute().map(_.list[GroupByIsAnonymous])
+val result: List[GroupByIsAnonymous] = Await.result(response, 10.seconds)
```
[the remark this answers is not in the document]
Comment (anskarl) [reply README.md]: I agree, I will remove the Awaits from all examples.

File: src/main/scala/ing/wbaa/druid/DruidQuery.scala (hunk 6 above)
Comment (anskarl) [reply src/main/scala/ing/wbaa/druid/DruidQuery.scala]: For some reason I haven't noticed that I didn't use capital when defining the object name in enumerations. You are right, Json serialization will convert it to lowercase.

**Comment by anskarl** [comment]
Hi @bjgbeelen 

Thank you for the code review! I have updated the PR with the requested changes (documentation and code). I have also made any required changes to resolve the conflicts with the latest version in master branch (after the merge of PR #37).

**Pull request merged**